pkg/webhook/resources/node/validator.go

```diff
@@ -0,0 +1,70 @@
+package node
+
+import (
+	v1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	ctlnode "github.com/harvester/harvester/pkg/controller/master/node"
+	werror "github.com/harvester/harvester/pkg/webhook/error"
+	"github.com/harvester/harvester/pkg/webhook/types"
+)
+
+func NewValidator(nodeCache v1.NodeCache) types.Validator {
+	return &nodeValidator{
+		nodeCache: nodeCache,
+	}
+}
+
+type nodeValidator struct {
+	types.DefaultValidator
+	nodeCache v1.NodeCache
+}
+
+func (v *nodeValidator) Resource() types.Resource {
+	return types.Resource{
+		Name:       "nodes",
+		Scope:      admissionregv1.ClusterScope,
+		APIGroup:   corev1.SchemeGroupVersion.Group,
+		APIVersion: corev1.SchemeGroupVersion.Version,
+		ObjectType: &corev1.Node{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Update,
+		},
+	}
+}
+
+func (v *nodeValidator) Update(request *types.Request, oldObj runtime.Object, newObj runtime.Object) error {
+	oldNode := oldObj.(*corev1.Node)
+	newNode := newObj.(*corev1.Node)
+
+	// if old node already have "maintain-status" annotation, it has already been enabled, so we skip it
+	if _, ok := oldNode.Annotations[ctlnode.MaintainStatusAnnotationKey]; ok {
+		return nil
+	}
+	// if new node doesn't have "maintain-status" annotation, we skip it
+	if _, ok := newNode.Annotations[ctlnode.MaintainStatusAnnotationKey]; !ok {
+		return nil
+	}
+
+	nodeList, err := v.nodeCache.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	nodeMaintainCount := 0
+	for _, node := range nodeList {
+		if _, ok := node.Annotations[ctlnode.MaintainStatusAnnotationKey]; !ok {
+			nodeMaintainCount++
```

Comment:
Instead of traversing all nodes, we can return once we find one available node.

Reply:
Updated it. Thank you.